src/meta/MetaHttpDownloadHandler.cpp

```diff
@@ -0,0 +1,236 @@
+    folly::split("\n", result.value(), files, true);
+    auto partNumber = std::count_if(files.begin(), files.end(), [](const std::string& element) {
+                                        return element.find(".sst") != std::string::npos;
+                                    });
+
+    std::vector<cpp2::HostItem> hostItems;
+    std::unique_ptr<kvstore::KVIterator> hostIter;
+    auto hostPrefix = MetaServiceUtils::hostPrefix();
+    auto hostRet = kvstore_->prefix(0, 0, hostPrefix, &hostIter);
+    if (hostRet != kvstore::ResultCode::SUCCEEDED) {
+        return false;
+    }
+
+    std::unordered_map<HostAddr, std::vector<PartitionID>> hostPartition;
+    while (hostIter->valid()) {
+        cpp2::HostItem item;
+        nebula::cpp2::HostAddr host;
+        auto hostAddrPiece = hostIter->key().subpiece(hostPrefix.size());
+        memcpy(&host, hostAddrPiece.data(), hostAddrPiece.size());
+        item.set_hostAddr(host);
+        auto address = std::make_pair(host.get_ip(), host.get_port());
+        auto hostIter_ = hostPartition.find(address);
+        if (hostIter_ == hostPartition.end()) {
+            std::vector<PartitionID> partitions;
+            hostPartition.insert(std::make_pair(address, partitions));
+        }
+        hostItems.emplace_back(item);
+        hostIter->next();
+    }
+
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto prefix = MetaServiceUtils::partPrefix(spaceID_);
+    auto ret = kvstore_->prefix(0, 0, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        return false;
+    }
+
+    int32_t partSize{0};
+    while (iter->valid()) {
+        auto key = iter->key();
+        PartitionID partId;
+        memcpy(&partId, key.data() + prefix.size(), sizeof(PartitionID));
+        for (auto host : MetaServiceUtils::parsePartVal(iter->val())) {
+            auto address = std::make_pair(host.get_ip(), host.get_port());
+            hostPartition[address].emplace_back(partId);
+        }
+        partSize++;
+        iter->next();
+    }
+
+    if (partNumber != partSize) {
+        LOG(ERROR) << "HDFS part number should be equal with nebula "
+                   << partNumber << " " << partSize;
+        return false;
+    }
+
+    std::atomic<int> completed(0);
+    std::vector<std::thread> threads;
+    for (auto &pair : hostPartition) {
+        std::vector<PartitionID> partitions;
+        for (auto part : pair.second) {
+            partitions.emplace_back(part - 1);
+        }
+        std::string partNumbers;
+        folly::join(",", partitions, partNumbers);
+
+        auto host = network::NetworkUtils::intToIPv4(pair.first.first);
+        auto t = "http://%s:%d/download?url=%s&port=%d&path=%s&parts=%s&local=%s";
+        auto download = folly::stringPrintf(t, host.c_str(), 22000, url.c_str(), port, path.c_str(),
+                                            partNumbers.c_str(), local.c_str());
+        command = folly::stringPrintf("curl %s ", download.c_str());
```

Comment:
Consider some cases where the url needs to be encoded. Please address it or add a TODO.

Reply:
because the url is composed by host, port and path.

I think it's in control .  Can you give me some examples about `encoded` ?